rust/arrow/src/bytes.rs

```diff
@@ -0,0 +1,185 @@
+/// foreign deallocator to deallocate the region when it is no longer needed.
+pub struct Bytes {
+    /// The raw pointer to be begining of the region
+    ptr: *const u8,
+
+    /// The number of bytes visible to this region. This is always smaller than its capacity (when avaliable).
+    len: usize,
+
+    /// how to deallocate this region
+    deallocation: Deallocation,
+}
+
+impl Bytes {
+    /// Takes ownership of an allocated memory region,
+    ///
+    /// # Arguments
+    ///
+    /// * `ptr` - Pointer to raw parts
+    /// * `len` - Length of raw parts in **bytes**
+    /// * `capacity` - Total allocated memory for the pointer `ptr`, in **bytes**
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe as there is no guarantee that the given pointer is valid for `len`
+    /// bytes. If the `ptr` and `capacity` come from a `Buffer`, then this is guaranteed.
+    pub unsafe fn new(ptr: *const u8, len: usize, deallocation: Deallocation) -> Bytes {
+        Bytes {
+            ptr,
+            len,
+            deallocation,
+        }
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe { slice::from_raw_parts(self.ptr, self.len) }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    #[inline]
+    pub fn raw_data(&self) -> *const u8 {
+        self.ptr
+    }
+
+    #[inline]
+    pub fn raw_data_mut(&mut self) -> *mut u8 {
+        self.ptr as *mut u8
+    }
+
+    pub fn capacity(&self) -> usize {
+        match self.deallocation {
+            Deallocation::Native(capacity) => capacity,
+            // we cannot determine this in general,
+            // and thus we state that this is externally-owned memory
+            Deallocation::Foreign(_) => 0,
+        }
+    }
+}
+
+impl Drop for Bytes {
+    #[inline]
+    fn drop(&mut self) {
+        match &self.deallocation {
+            Deallocation::Native(capacity) => {
+                if !self.ptr.is_null() {
+                    unsafe { memory::free_aligned(self.ptr as *mut u8, *capacity) };
+                }
+            }
+            Deallocation::Foreign(drop) => {
+                (drop.clone())(self);
+            }
+        }
+    }
+}
+
+impl PartialEq for Bytes {
+    fn eq(&self, other: &Bytes) -> bool {
+        self.as_slice() == other.as_slice()
+    }
+}
+
+impl Debug for Bytes {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "Bytes {{ ptr: {:?}, len: {}, data: ", self.ptr, self.len,)?;
+
+        f.debug_list().entries(self.as_slice().iter()).finish()?;
+
+        write!(f, " }}")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_dealloc() {
+        let a = Box::new(b"hello");
+
+        let dealloc = Arc::new(|bytes: &mut Bytes| {
+            // println!(""); seems to be the only way to validate that this is actually called
```

Comment:
Perhaps using a closure that mutates a captured variable?
Apparently you need to use either `FnMut` or a `Cell`:
https://stackoverflow.com/questions/38677736/passing-a-closure-that-modifies-its-environment-to-a-function-in-rust

Reply:
Done. I had to introduce a `Mutex`, because a `FnMut` is now mutable, while the data itself is not. I am not very happy with this, but it makes sense as we cannot assume that the C data interface is thread-safe, right?

Reply:
I reverted this. I do not think that that `Fn` should be mutable, as it is just performing an FFI call, over which Rust does not need to know about mutability. I am still trying to test it, but I think that something like `Arc<dyn Fn(&mut Bytes)>` is a better signature,.

Reply:
Why don't you use a `Cell`?

Reply:
```rust
        let b = Cell::new(false);
        let dealloc = Arc::new(|bytes: &mut Bytes| {
            *b.get_mut() = true;
            assert_eq!(bytes.as_slice(), &b"hello"[1..4]);
        });
```

does not compile because it requires moving `b` to inside the closure: if we move `b` to inside the closure (using `move`), the closure is no longer `Fn`, but `FnMut`. If the closure is `FnMut`, we can no longer wrap it inside an `Arc`, as `Arc` is immutable. To make it immutable, we need to wrap it around a `Mutex`.

The difference between the code we are going here and the example in SO is that we have an immutable function, as the underlying resource that this function acts upon is outside rust. I.e. from rust's perspective, the function is `Fn`.

One way to test this would be to make the closure to write something to a file, and verify that that was written. I.e. test that the function mutated something outside of Rust.

Reply:
After trying out varying things, I got the following to work:
```rust
use std::cell::Cell;
use std::sync::Arc;

pub type VoidFn = Arc<dyn Fn()>;

fn main() {
    let integer = Arc::new(Cell::new(5));
    let inner = integer.clone();
    let closure = Arc::new(move || {
        inner.set(inner.get() + 1);
    });
    execute_closure(closure);
    println!("After closure: {}", integer.get());
}

fn execute_closure(func: VoidFn)
{
    func();
}
```

I may be missing something though.

Reply:
By the way, by definition a destructor will mutate some state (visible or not), so it seems `FnMut` may be fine too.

Reply:
That is awesome! Thanks a lot for the help and insight @pitrou .

The destructor mutates state. Shouldn't that state be only the state of `self`? IMO that is the reason rust's `Drop` requires a mutable ref of self, `fn drop(&mut self)`.

Reply:
I have no idea, I am out of my depth here (not a Rust developer :-)).